ASoC: SOF: add ipc4 support

bardliao opened the pull request:
We plan to support a new IPC version named IPC4 on Intel next platforms.
Compared to current IPC3, the key difference are:
(1) pipeline is composed of module, not component. And also pipeline will
    manage memory allocation including buffer
(2) there is no buffer defined in topology and FW will create buffer on
    demand

The pipeline in IPC4 has different operations and states but it can be
converted to current IPC3 straight forward. Module interface can also be
supported by current component interface.

plbossart commented on include/sound/sof/header2.h:
```diff
@@ -0,0 +1,410 @@
+	PIPE_PAUSED = 3,
+	PIPE_RUNNING = 4,
+	PIPE_EOS = 5
+};
```
you will need prefixes for all the enum definitions, e.g. sof_ipc_pipeline_state, SOF_IPC_PIPE_RESET

What you have for now is way too generic and conflict prone.

bardliao replied:
ack

plbossart commented on include/sound/sof/header2.h:
```diff
@@ -0,0 +1,410 @@
+	CHANNEL_CONFIG_I2S_DUAL_STEREO_1	= 11,
+	/* L, C, R, Ls, Rs & LFE., LS, RS */
+	CHANNEL_CONFIG_7_POINT_1		= 12,
+	CHANNEL_CONFIG_INVALID
```
what on earth is this about?

Again prefixes needed!

bardliao replied:
ack

plbossart commented on include/sound/sof/header2.h:
```diff
@@ -0,0 +1,410 @@
+	uint32_t channels_count : 8;
+	uint32_t valid_bit_depth : 8;
+	enum sample_type s_type : 8;
+	uint32_t reserved : 8;
```
NAK, no bitfields in IPC definitions.

bardliao replied:
> NAK, no bitfields in IPC definitions.

I have to figure a better way to do this as bitfields is used on the firmware side.

plbossart commented on include/sound/sof/header2.h:
```diff
@@ -0,0 +1,410 @@
+	/* SNDW link output (DSP ->) */
+	kAlhSndWireStreamLinkOutputClass = 16,
+	/* SNDW link input (DSP <-) */
+	kAlhSndWireStreamLinkInputClass = 17,
```
why do we have two definitions for 16 and 17?

bardliao replied:
> why do we have two definitions for 16 and 17?

That was copied from cavs firmware and I don't know the reason why they use the same number. @RanderWang Do you know any reason that we have two definitions for 16 and 17?

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,474 @@
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2018 Intel Corporation. All rights reserved.
```
2021

bardliao replied:
ack

plbossart commented on sound/soc/sof/Makefile:
```diff
@@ -1,7 +1,7 @@
 
 snd-sof-objs := core.o ops.o loader.o ipc.o pcm.o pm.o debug.o topology.o\
-		control.o trace.o utils.o sof-audio.o
+		control.o trace.o utils.o sof-audio.o ipc4.o
```
do we need both ipc and ipc4 in the same module?

bardliao replied:
> do we need both ipc and ipc4 in the same module?

Yes, there are still some common functions in ipc.c. For example, snd_sof_ipc_init and snd_sof_ipc_free.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,474 @@
+#include "ipc4.h"
+#include "ops.h"
+
+static struct sof_ipc4_fw_status ipc4_status[] = {
```
const?

bardliao replied:
ack

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,474 @@
+
+	sdev->ipc->msg.reply_error = status;
+	if (status) {
+		for (i = 0; i < sizeof(ipc4_status) / sizeof(struct sof_ipc4_fw_status); i++) {
```
ARRAY_SIZE?

bardliao replied:
ack

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,474 @@
+{
+	int *data = (int *)config;
+	int i, dw_size;
+	char buffer[36];
```
magic number!

bardliao replied:
> magic number!

@RanderWang Could you help on it? Where does 36 from?

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,474 @@
+
+	dw_size = size >> 2;
+	for (i = 0; i < dw_size; i += 4) {
+		int len = (i + 4 < dw_size) ? 16 : (dw_size - i) * 4;
```
this is horrible, please simplify

dbaluta commented on include/sound/sof/header2.h:
```diff
@@ -0,0 +1,410 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause) */
+/*
```
Can you make a little  bit more clear what is IPC4 introduced for?

- so it's a new version of IPC
- it is needed because we need to make the IPC compatible with Windows implemenation (?)


header2.h naming is very strange naming given that ipc is called ipc4.

bardliao replied:
Thanks @dbaluta. Moved the header2.h to ipc4/header.h and updated the commit message. I will work with my colleagues to improve the commit message.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,474 @@
+	/* attach any data */
+	if (msg_bytes > msg->msg_data_size) {
+		devm_kfree(sdev->dev, msg->msg_data);
+		msg->msg_data = devm_kzalloc(sdev->dev, msg_bytes, GFP_KERNEL);
```
no. don't use devm_ for stuff like this.

bardliao replied:
@plbossart I will remove the reallocate stuff from this PR and we will create another PR to allocate enough space at beginning.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,474 @@
+	u32 header;
+	int ret;
+
+	dev_dbg(sdev->dev, "ipc4 set pipeline %d status %d", id, status);
```
way too many dev_dbg() that's not going to fly when we upstream

plbossart commented on sound/soc/sof/ipc4.h:
```diff
@@ -0,0 +1,42 @@
+static inline int sof_ipc4_module_buffer_size(int ch,  int rate, int width, int sch_num)
+{
+	/* 1ms data size in fw */
+	return (ch) * ((rate) / 1000) * ((width) >> 3) * sch_num;
```
that's broken for 44.1 !

bardliao replied:
Indeed, it is not accurate. I don't know how should we handle this. Say if e.g. ch=2, rate=44100, width=16, sch_num=2. The accurate value is 352.8, but buffer size is an integer and will be 352. @RanderWang What do you think?

RanderWang replied:
yes, we need to allocated buffer with rate of 48k for 44.1k. Now we don't support 44.1 in FW and we need to convert it to 48k. And even we can support native 44.1, a little bigger buffer is no harmful.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -383,6 +387,17 @@ enum snd_sof_fw_state {
+	u32	*instance_id; /* bit array */
+	u16 instance_max_count;
+	u32 bss_size;
+	void *private;
```
align to a tab or keep a single space, don't add random spaces.

bardliao replied:
ack

plbossart reviewed (changes requested):
This needs a lot of cleanups @bardliao @lgirdwood @RanderWang @ranj063 

The header files seem to be a copy/paste, it needs proper prefixes, comments.

bardliao commented:
@plbossart Thanks for your review. I addressed most of them. To make the PR smaller and easier to remove, I removed the module and pipeline stuff. I will create separate PRs for them after this PR is merged. The major question to me is that how should we handle the bitfields while it is used on the firmware side. Also I am wondering should we use dual-license for ipc4?

plbossart commented:
> @plbossart Thanks for your review. I addressed most of them. To make the PR smaller and easier to remove, I removed the module and pipeline stuff. I will create separate PRs for them after this PR is merged. The major question to me is that how should we handle the bitfields while it is used on the firmware side. Also I am wondering should we use dual-license for ipc4?

I think the dual license applies here, this is extracted from BSD stuff on the firmware side and we want to make it usable in a Linux context as well as other OSes if they so desired.

RanderWang commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,307 @@
+	SOF_IPC4_GLB_NOTIFY_RESOURCE_EVENT = 5,
+	SOF_IPC4_GLB_NOTIFY_LOG_BUFFER_STATUS = 6,
+	SOF_IPC4_GLB_NOTIFY_TIMESTAMP_CAPTURED = 7,
+	SOF_IPC4_GLB_NOTIFY_FW_READY = 8
```
@bardliao we can add item 9 to item max

bardliao replied:
> @bardliao we can add item 9 to item max

Sorry @RanderWang I don't see where will the MAX item be used. Could you point out where will it be used? Thanks.

Edit: I got your point and added all notification types now.

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,324 @@
+};
+
+enum sof_ipc4_pipe_line_state {
+	SOF_IPC4_PIPE_RESET = 2,
```
maybe add a comment on why 0 and 1 are skipped?

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,324 @@
+	SOF_IPC4_FS_88200HZ	= 88200, /* AAC, SRC only. */
+	SOF_IPC4_FS_96000HZ	= 96000, /* AAC, SRC only. */
+	SOF_IPC4_FS_176400HZ	= 176400, /* SRC only. */
+	SOF_IPC4_FS_192000HZ	= 192000, /* SRC only. */
```
can we remove all these AAC, MP3, SRC reference?

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,324 @@
+	SOF_IPC4_FS_96000HZ	= 96000, /* AAC, SRC only. */
+	SOF_IPC4_FS_176400HZ	= 176400, /* SRC only. */
+	SOF_IPC4_FS_192000HZ	= 192000, /* SRC only. */
+	SOF_IPC4_FS_INVALID
```
so the value of 192001 is invalid? This should be an explicit value really, it makes no sense to mix explicit values and implicit increments.

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,324 @@
+	SOF_IPC4_DEPTH_24BIT	= 24, /* 24 bits depth - Default */
+	SOF_IPC4_DEPTH_32BIT	= 32, /* 32 bits depth */
+	SOF_IPC4_DEPTH_64BIT	= 64, /* 64 bits depth */
+	SOF_IPC4_DEPTH_INVALID
```
same here, this would mean 65 is the invalid depth?

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,324 @@
+	SOF_IPC4_CHANNEL_CONFIG_I2S_DUAL_STEREO_1	= 11,
+	/* L, C, R, Ls, Rs & LFE., LS, RS */
+	SOF_IPC4_CHANNEL_CONFIG_7_POINT_1		= 12,
+	SOF_IPC4_CHANNEL_CONFIG_INVALID
```
same here, 13 is invalid?

bardliao replied:
> same here, 13 is invalid?

Those invalid enum values are copied from the existing cavs firmware. It will be mismatched if we change it. The question is that do we need to be compatible with the existing cavs firmware? Or be compatible with the converged firmware is good enough?  @plbossart @lgirdwood

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,324 @@
+#define SOF_IPC4_LSB_INTEGER		1, /* integer with Least Significant Byte first */
+#define SOF_IPC4_SIGNED_INTEGER		2, /* signed integer */
+#define SOF_IPC4_UNSIGNED_INTEGER	3, /* unsigned integer */
+#define SOF_IPC4_FLOAT			4 /* unsigned integer */
```
can this be an enum?

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,324 @@
+	/* Exception caught by DSP FW */
+	SOF_IPC4_GLB_NOTIFY_EXCEPTION_CAUGHT = 10,
+	/* Custom module notification */
+	SOF_IPC4_GLB_NOTIFY_MODULE_NOTIFICATION = 12,
```
add a comment on 11 being skipped

bardliao replied:
> add a comment on 11 being skipped

That was copied from the existing cavs firmware. I have no idea why it starts with 4 and skip 11. Is `"To be compatible with the existing cavs firmware"` ok?

plbossart reviewed (changes requested):
I know the definitions came from our firmware friends, but I am afraid this is still questionable @bardliao @lgirdwood

bardliao commented:
> I know the definitions came from our firmware friends, but I am afraid this is still questionable @bardliao @lgirdwood

@plbossart I addressed most of your comments. However, I still have compatibility concern for the _INVALID value in each enum. Do we need to be compatible with the existing cavs firmware? Or be compatible with the converged firmware is good enough? @lgirdwood

bardliao commented:
> > I know the definitions came from our firmware friends, but I am afraid this is still questionable @bardliao @lgirdwood
> 
> @plbossart I addressed most of your comments. However, I still have compatibility concern for the _INVALID value in each enum. Do we need to be compatible with the existing cavs firmware? Or be compatible with the converged firmware is good enough? @lgirdwood

I removed all the _INVALID definitions and we can avoid using it on kernel side.

plbossart commented:
Sound good to me @bardliao and @lgirdwood, but is there anything related to ABI here? How would we identify a platform as requiring the use of IPCv3 v. IPCv4?

bardliao commented:
> Sound good to me @bardliao and @lgirdwood, but is there anything related to ABI here? How would we identify a platform as requiring the use of IPCv3 v. IPCv4?

I think there is nothing related to ABI. We will use an ops in `struct snd_sof_dsp_ops {}` to identify which IPC version will a platform use. See the new commit in this PR.

plbossart commented:
> > Sound good to me @bardliao and @lgirdwood, but is there anything related to ABI here? How would we identify a platform as requiring the use of IPCv3 v. IPCv4?
> 
> I think there is nothing related to ABI. We will use an ops in `struct snd_sof_dsp_ops {}` to identify which IPC version will a platform use. See the new commit in this PR.

But even if you know IPC4 is used for a given platform, how would you track ABI changes for that IPC4? Is this using the same versioning number as IPC3, or is this a different accounting altogether? @lgirdwood it's important we get this right.

lgirdwood commented:
> But even if you know IPC4 is used for a given platform, how would you track ABI changes for that IPC4? Is this using the same versioning number as IPC3, or is this a different accounting altogether? @lgirdwood it's important we get this right.

The same methods as IPC3 - Ideally this will be short term until we can move to IPC tuples.

plbossart commented:
@lgirdwood the two types of structures are going to live in parallel. If we need an update to an IPC3 structure, how would we track it?
If we move to ABI 4.0 when IPC4 is introduced, would a change to an IPC3 structure be tracked as 4.1? or 3.20?

lgirdwood commented:
> @lgirdwood the two types of structures are going to live in parallel. If we need an update to an IPC3 structure, how would we track it?
> If we move to ABI 4.0 when IPC4 is introduced, would a change to an IPC3 structure be tracked as 4.1? or 3.20?

IPC 3 changes would be tracked as IPC 3.20 and so on. Likewise IPC4 updates would be IPC4.1 and so on. We should be able to get rid of the tracking for most data elems with IPC tuples.

bardliao commented:
> > @lgirdwood the two types of structures are going to live in parallel. If we need an update to an IPC3 structure, how would we track it?
> > If we move to ABI 4.0 when IPC4 is introduced, would a change to an IPC3 structure be tracked as 4.1? or 3.20?
> 
> IPC 3 changes would be tracked as IPC 3.20 and so on. Likewise IPC4 updates would be IPC4.1 and so on. We should be able to get rid of the tracking for most data elems with IPC tuples.

@lgirdwood So there will be two ABI versions one for IPC3 and the other for IPC4, right? And if we change something irrelevant to IPC, both ABI versions will be increased. Is my understanding correct?

ujfalusi commented on sound/soc/sof/ops.h:
```diff
@@ -348,6 +348,14 @@ static inline int snd_sof_dsp_send_msg(struct snd_sof_dev *sdev,
+	if (sof_ops(sdev)->get_ipc_version)
+		return sof_ops(sdev)->get_ipc_version(sdev);
+
+	return SOF_IPC_VERSION_1;
```
The commit message refers to `SOF_INTEL_IPC_VERSION_1`
which I assume is IPC3 ?

bardliao replied:
> The commit message refers to `SOF_INTEL_IPC_VERSION_1`
> which I assume is IPC3 ?

Yes, SOF_IPC_VERSION_1 is IPC3 and SOF_IPC_VERSION_2 is IPC4

ujfalusi replied:
Would it not be confusing if we refer to the IPC3 as `SOF_IPC_VERSION_3` and IPC4 as `SOF_IPC_VERSION_4`?
But based on your other comment, the IPC4 is a a version used by closed-source firmware? Is it valid to say `SOF_IPC_VERSION_x` to a protocol which is _not_ used by SOF?

bardliao replied:
> Would it not be confusing if we refer to the IPC3 as `SOF_IPC_VERSION_3` and IPC4 as `SOF_IPC_VERSION_4`?

That is a good idea indeed. The SOF_IPC_VERSION_x is implemented before we name the new IPC as IPC4 and the current IPC as IPC3. That's why it is SOF_IPC_VERSION_1 and 2.

> But based on your other comment, the IPC4 is a a version used by closed-source firmware? Is it valid to say `SOF_IPC_VERSION_x` to a protocol which is _not_ used by SOF?

IPC4 is used by closed-source firmware and Windows driver, yes.

ujfalusi commented on include/sound/sof/ipc4/header.h:
```diff
@@ -0,0 +1,315 @@
+ * IPC4 messages have two 32 bit identifier made up as follows :-
+ *
+ * header - msg type, msg id, msg direction ...
+ * extension - extra params such as msg data size in mailbox
```
and these 'identifiers' are not part of the message (msg_data) itself which is placed in mailbox?

bardliao replied:
> and these 'identifiers' are not part of the message (msg_data) itself which is placed in mailbox?

They are part of `snd_sof_ipc_msg`, but not part of `msg_data`.

ujfalusi commented on sound/soc/sof/sof-priv.h:
```diff
@@ -353,7 +353,9 @@ struct snd_sof_mailbox {
 	u32 header;
+	u32 extension;
 	void *msg_data;
+	int msg_data_size;
```
How is this relates to msg_size?

bardliao replied:
> How is this relates to msg_size?

msg_data_size is not related to msg_size. It indicates the size of msg_data. We currently allocate msg_data to a fixed SOF_IPC_MSG_MAX_SIZE. We will allocate a different size for different platform.

kv2019i commented on sound/soc/sof/sof-priv.h:
```diff
@@ -74,6 +74,12 @@ extern int sof_core_debug;
 /* max number of DSP cores */
 #define SOF_MAX_DSP_NUM_CORES 8
 
+enum sof_intel_ipc_version {
```
I don't think this can be "Intel IPC" version as the IPC is common.

And I'm not sure it's a good to introduce a concept of "IPC version" that is different from SOF ABI major version.  The SOF ABI really is a version of the IPC, so this is a duplicate definition.

ujfalusi commented on sound/soc/sof/ipc4.c:
```diff
@@ -0,0 +1,221 @@
+
+	sdev->msg = msg;
+
+	ret = snd_sof_dsp_send_msg(sdev, msg);
```
The platform code will assign either ipc4 or ipc3 version of `send_msg` ops, right?
Is it a possibility that a module would try to send IPC3 message while the platform set the IPC4 callbacks?
I don't think we have any, but in theory there could be external third party modules tapping in to SOF or if we miss some IPC message updates...

bardliao replied:
> The platform code will assign either ipc4 or ipc3 version of `send_msg` ops, right?
> Is it a possibility that a module would try to send IPC3 message while the platform set the IPC4 callbacks?

Yes, and the we will fail to get the firmware ready IPC. I tend to agree that using ABI MAJOR version which is provided by firmware itself is a more reasonable solution.

> I don't think we have any, but in theory there could be external third party modules tapping in to SOF or if we miss some IPC message updates...

ujfalusi reviewed (changes requested):
Do you know if we are going to support dma-trace, probes with IPC4? How does the change going to look like?

kv2019i reviewed (changes requested):
My take on the ABI versioning: I'd propose we:
- use ABI major component as reported by FW as the IPC version (either 3 or 4 currrently)
    - existing FW_READY message (in FW manifest) can relay the version used, no new interface needed and information is known before IPC is used
- 3.x and 4.x versions continue to be maintained with their own MINOR versions
- SOF FW sources in main branch support both 3.x and 4.x
    - but not necessarily for all HW targets, some may support both, some 3.x, some only 4.x
- a single loaded SOF FW binary only supports 3.x or 4.x, not both
- kernel driver will support both 3.x and 4.x FWs and will switch IPC variant based on the version used by the loaded FW
     - version checks are only needed in places where the two variants are actually different

bardliao commented:
> My take on the ABI versioning: I'd propose we:
> 
> * use ABI major component as reported by FW as the IPC version (either 3 or 4 currrently)
>   
>   * existing FW_READY message (in FW manifest) can relay the version used, no new interface needed and information is known before IPC is used
> * 3.x and 4.x versions continue to be maintained with their own MINOR versions
> * SOF FW sources in main branch support both 3.x and 4.x
>   
>   * but not necessarily for all HW targets, some may support both, some 3.x, some only 4.x
> * a single loaded SOF FW binary only supports 3.x or 4.x, not both
> * kernel driver will support both 3.x and 4.x FWs and will switch IPC variant based on the version used by the loaded FW
>   
>   * version checks are only needed in places where the two variants are actually different

Thanks @kv2019i for the propose. However, it will not be compatible with the existing closed-source firmware. So, the question is that do we need to be compatible with the existing closed-source firmware?

bardliao commented:
> Do you know if we are going to support dma-trace, probes with IPC4? How does the change going to look like?

I think this question is for @RanderWang :)
